drivers/w3s.go

```diff
@@ -0,0 +1,340 @@
+}
+
+type W3sSession struct {
+	os *W3sOS
+}
+
+func NewW3sDriver(ucanKey, ucanProof, dirPath, pubId string) *W3sOS {
+	return &W3sOS{
+		ucanKey:   ucanKey,
+		ucanProof: ucanProof,
+		dirPath:   dirPath,
+		pubId:     pubId,
+	}
+}
+
+func (ostore *W3sOS) NewSession(filename string) OSSession {
+	if filename != "" {
+		return nil
+	}
+	session := &W3sSession{
+		os: ostore,
+	}
+	return session
+}
+
+func (ostore *W3sOS) UriSchemes() []string {
+	return []string{}
+}
+
+func (ostore *W3sOS) Description() string {
+	return "Web3 Storage driver."
+}
+
+func (session *W3sSession) OS() OSDriver {
+	return session.os
+}
+
+func (session *W3sSession) EndSession() {
+	// no op
+}
+
+func (session *W3sSession) ListFiles(ctx context.Context, cid, delim string) (PageInfo, error) {
+	return nil, ErrNotSupported
+}
+
+func (session *W3sSession) ReadData(ctx context.Context, name string) (*FileInfoReader, error) {
+	return nil, ErrNotSupported
+}
+
+func (session *W3sSession) IsExternal() bool {
+	return false
+}
+
+func (session *W3sSession) IsOwn(url string) bool {
+	// not used anywhere
+	return false
+}
+
+func (session *W3sSession) GetInfo() *OSInfo {
+	return nil
+}
+
+func (session *W3sSession) getAbsolutePath(name string) string {
+	// not supported
+	return ""
+}
+
+func (session *W3sSession) SaveData(ctx context.Context, name string, data io.Reader, meta map[string]string, timeout time.Duration) (string, error) {
+	if timeout <= 0 {
+		timeout = w3SDefaultSaveTimeout
+	}
+	ctx, cancel := context.WithTimeout(ctx, timeout)
+	defer cancel()
+
+	filePath, err := toFile(data)
+	if err != nil {
+		return "", err
+	}
+	defer deleteFile(filePath)
+
+	carPath, fileCid, err := ipfsCarPack(ctx, filePath)
+	if err != nil {
+		return "", err
+	}
+	defer deleteFile(carPath)
+
+	carCid, err := w3StoreCar(ctx, carPath)
+	if err != nil {
+		return "", err
+	}
+
+	rCar := session.os.getRootCar()
+	if err = rCar.addFile(ctx, session.os.dirPath, name, fileCid, carCid); err != nil {
+		return "", err
+	}
+
+	return fileCid, nil
+}
+
+func (rc *rootCar) addFile(ctx context.Context, dirPath, filename, fileCid, carCid string) error {
+	rc.mu.Lock()
+	defer rc.mu.Unlock()
+
+	rc.carCids = append(rc.carCids, carCid)
+
+	// split path by "/", ignore empty strings
+	dirPaths := strings.FieldsFunc(dirPath, func(c rune) bool { return c == '/' })
+
+	newRoot, err := rc.addFileToDag(ctx, rc.root, dirPaths, filename, fileCid)
+	if err != nil {
+		return err
+	}
+	rc.root = newRoot
+
+	return nil
+}
+
+func (rc *rootCar) addFileToDag(ctx context.Context, n *merkledag.ProtoNode, dirPaths []string, filename, fileCid string) (*merkledag.ProtoNode, error) {
+	if len(dirPaths) == 0 {
+		// n is a leaf
+		fCid, err := cid.Parse(fileCid)
+		if err != nil {
+			return nil, err
+		}
+		n.AddRawLink(filename, &format.Link{Cid: fCid})
+		rc.dag.Add(ctx, n)
+		return n, nil
+	}
+
+	// n is not a leaf, recursively update until the leaf
+	head, tail := dirPaths[0], dirPaths[1:]
```

Comment:
```suggestion
	rootPath, childPaths := dirPaths[0], dirPaths[1:]
```

This might help with readability of the logic (which is complex) so that the reader can distinguish between vars that are paths vs. vars that are of type `merkledag.ProtoNode`. Personally, I had some trouble parsing through the head, tail, child terminology.